packages/nitro-protocol/contracts/MultiAssetHolder.sol

```diff
@@ -2,104 +2,220 @@
+            assetOutcome.allocationOrGuaranteeBytes,
+            (Outcome.AllocationItem[])
+        );
+        address asset = outcome[assetIndex].assetHolderAddress;
+
         // effects and interactions
-        _transfer(fromChannelId, allocationBytes, indices);
+        allocation = _transfer(asset, fromChannelId, allocation, indices); // update in place to newAllocation
+        outcome[assetIndex].assetOutcomeBytes = abi.encode(
+            Outcome.AssetOutcome(Outcome.AssetOutcomeType.Allocation, abi.encode(allocation))
+        );
+        bytes32 outcomeHash = keccak256(abi.encode(outcome));
+        _updateFingerprint(fromChannelId, stateHash, challengerAddress, outcomeHash);
+        emit FingerprintUpdated(fromChannelId, abi.encode(outcome));
+        return abi.encode(outcome);
     }
 
     /**
      * @notice Transfers as many funds escrowed against `guarantorChannelId` as can be afforded for a specific destination in the beneficiaries of the __target__ of that channel. Checks against the storage in this contract.
      * @dev Transfers as many funds escrowed against `guarantorChannelId` as can be afforded for a specific destination in the beneficiaries of the __target__ of that channel. Checks against the storage in this contract.
-     * @param guarantorChannelId Unique identifier for a guarantor state channel.
-     * @param guaranteeBytes The abi.encode of Outcome.Guarantee
-     * @param allocationBytes The abi.encode of AssetOutcome.Allocation for the __target__
-     * @param indices Array with each entry denoting the index of a destination (in the target channel) to transfer funds to. Should be in increasing order. An empty array indicates "all".
      */
     function claim(
+        uint256 assetIndex, // TODO consider a uint48?
         bytes32 guarantorChannelId,
-        bytes calldata guaranteeBytes,
-        bytes calldata allocationBytes,
+        bytes memory guarantorOutcomeBytes,
+        bytes32 guarantorStateHash,
+        address guarantorChallengerAddress,
+        bytes memory targetOutcomeBytes,
+        bytes32 targetStateHash,
+        address targetChallengerAddress,
         uint256[] memory indices
-    ) external override {
+    ) public override {
         // checks
-        _requireIncreasingIndices(indices);
-        _requireCorrectGuaranteeHash(guarantorChannelId, guaranteeBytes);
-        Outcome.Guarantee memory guarantee = abi.decode(guaranteeBytes, (Outcome.Guarantee));
-        _requireCorrectAllocationHash(guarantee.targetChannelId, allocationBytes);
-        // effects and interactions
-        _claim(guarantorChannelId, guarantee, allocationBytes, indices);
-    }
-
-    // **************
-    // Permissioned methods
-    // **************
-
-    modifier AdjudicatorOnly {
-        require(msg.sender == AdjudicatorAddress, 'Only NitroAdjudicator authorized');
-        _;
-    }
+        Outcome.Guarantee memory guarantee;
+        address asset;
+        {
```

Comment:
Curious, what is the reason there are separate blocks in this function?

Reply:
They are there to prevent "stack too deep" problems. See point 2 of this page https://soliditydeveloper.com/stacktoodeep .